### DISCUSSION TITLE
When will ESP32 ADC2 "bug" be resolved ?

### DISCUSSION THREAD

**Discussion by cacengineering** [arendst/Tasmota#19098] [Q&A] [answered]
We are developing around ESP32-32 series chipsets, and using ESP32_WROOM-32 devkits.  We took this approach as need multiple ADC inputs including a fast interrupt based one.  The ESP32 multiplexes the pins to the ADC's. The plan was to use fast interrupt on ADC1 and the other ADC inputs on ADC2.  However on reading of posts ADC2 is disabled in Tasmota due to conflicts with WiFi ?  Unfortunately, ADC1 does not have a full complement of ADC pins anyway (ADC1_1 and ADC1_2 don't exist in the architecture of the devkits) so not possible to only rely on ADC1.  We can try slow multiplex some of the pins in the hardware but ideal would be to have Tasmota implement ADC2.  Is this complicated and any idea of timelines if it is planned ?  Any other workarounds the community can suggest?

**Comment by sfromis** [comment]
It is a hardware limitation of the ESP32 that ADC2 does not work with Wifi, not a bug. Tasmota cannot do anything about that.

**Reply by cacengineering** [reply]
OK - so just to confirm - not a Tasmota specific issue ?  Are there any other Expressif chipsets that don't have this problem ?

**Reply by sfromis** [reply]
Not worked with this, but on ESP32-S2 and S3, it seems that ADC2 is not unavailable with Wifi, but may fail with Wifi having higher priority. Still, Tasmota has no support for more than 8 channels on ADC1.

**Comment by tablatronix** [comment]
There are ways to mitigate these issues with ADC2, not sure if tasmota uses any of them, but either way you cannot get the full channel or more than a couple reads, without sleeping wifi afaik, and adcs are only useful if oversampling most of the time especially on the ESP32 with its crappy adcs.

You should really consider solutions and tests in the ESP lib and SDK resources and not tasmota though, but I don't recall seeing anything groundbreaking. I am dealing with this right now, I have voltage dividers on adc2 pins, to use another pin would require a pcb respin

**Reply by cacengineering** [reply]
Much appreciate the input.  Am trying to stick with Tasmota if we can - does a lot of the stuff out of the box if we can get the configuration right - and understand the shortcomings ..... While we have some fast changing ADC channels most are slow and we will now use smoothing capacitors so slow sampling should be ok - even if we have to switch between inputs to overcome the limited ADC ports.  By the way - we are testing zener based voltage offsets instead of voltage dividers as a way of getting full range to the ADC and mitigating poor sampling.  We have accepted we need to respin the board :( and are now extensively breadboarding and testing before committing - how we picked up the ADC issue.  Might still need to go ESP lib ... or look into the Tasmota code but right now need to get a mostly working demo out ...

**Comment by tablatronix** [comment]
Yeah I think ANY solution would be have to be rolled into a tasmota fork as eperimental, I have not looked for discussion on this in esp32 github or espressif forums, but I am sure there is plenty discussion on this elsewhere

**Comment by Jason2866** [comment, marked as answer]
Most probably we will not add code to use the crappy ADC2. One thing is sure we will not invest time in this.

**Reply by cacengineering** [reply]
Thaks for the feedback.  Might be worth putting a note up on the Wiki highlighting ESP32 ADC2 has series limitations and Tasmota have no plans to implement.

**Reply by Jason2866** [reply]
We would have a lot of infos littering the documentation which is already big.
It is documented what's working. Not described stuff is not implemented. That's it.

**Comment by tablatronix** [comment]
https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/peripherals/adc.html#adc-limitations

It is even output in debug logging in esp32 when you use these pins
```log_e("GPIO%u: %s: ADC2 is in use by Wi-Fi. Please se```

**Comment by tablatronix** [comment]
Anyone roll their own ADC2 reads? the ESP32-CAM boards only expose ADC2, I am working on making sure I can read at least one channel

**Reply by sfromis** [reply]
Ask Espressif if they want to do something about their documented restriction about ADC2 not working well with Wifi. My guess is to not hold your breath waiting.

**Reply by Jason2866** [reply]
Add a extern (better) ADC chip. Better get a S3 cam board ;-)
